pinot-plugins/pinot-input-format/pinot-orc/src/main/java/org/apache/pinot/plugin/inputformat/orc/ORCRecordReader.java

```diff
@@ -95,25 +171,151 @@ public GenericRow next()
   @Override
   public GenericRow next(GenericRow reuse)
       throws IOException {
-    _recordReader.nextBatch(_reusableVectorizedRowBatch);
-    return _recordExtractor.extract(_reusableVectorizedRowBatch, reuse);
+    int numFields = _orcFields.size();
```

Comment:
Hive's ORC reader has a zero copy option to boost performance when reading from HDFS -- https://issues.apache.org/jira/browse/HIVE-6347

Secondly, I wonder if we can do something about making this vectorized. Probably not because we anyway have to create a GenericRow and return it for every next() call. However, it is possible to read a single column vector at a time from VectorizedRowBatch and avoid the repeated dynamic dispatch in the loop

Say if we do vectorized reads from each column vector from a single VectorizedRowBatch and then pick values from each cell to create a batch of GenericRow, would performance be any different? In that case, the next() API semantics can be changed to bulk based.

Reply:
The fact that we have to materialize the data read immediately into row loses any benefit we could have got from vectorization. May be if we do this in bulk manner, it could change things.

Reply:
This is not something worth optimizing - unless we can save hours. This code is not used in the query path.